pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -1843,6 +1842,27 @@ impl ResidentLayer {
         }
     }
 
+    /// Read all they keys in this layer in the given range, and write them all to
+    /// the provided writer.  Return the number of keys written.
+    #[tracing::instrument(level = tracing::Level::DEBUG, skip_all, fields(layer=%self))]
+    pub(crate) async fn rewrite<'a, F>(
+        &'a self,
+        range: Range<Key>,
+        filter: F,
```

Comment:
Do we have some need for different kind of filters?

Ok, it's for not hoisting the shard key selection up here. Unsure do we need these low level ops to be generic, but here I guess it doesn't hurt here.

Rewrite could be mean anything and the naming direction is at least different than of mine with `copy_lsn_prefix`, which is however quite similar to what this is doing, the generic filter fn for `copy_delta_prefix` is just `|_k, lsn| lsn <= until`, but I wanted to do it non-generic.

Reply:
I could imagine using this for something arbitrary like breaking up some oversized layer into smaller ones, but for the moment the ShardIdentity is the only kind of filtering we care about.  I've made this a bit less genericc to just take a ShardIdentity instead of an arbitrary function